### PR TITLE
Update Xcode parsing lib

### PR DESCRIPTION
**Pull request by lpusok**
### Changes
* Updated iOS Scanner to use absolute paths everywhere, in preparation of the updated xcodeproj package. This required updating a few Cocoapods parsing Ruby scripts.
* Fixed/worked around a problem that only top-level (*Pod* target) projects were parsed from Podfiles. Now using the first project, disregarding which target is it in.
```ruby
# Uncomment the next line to define a global platform for your project
# platform :ios, '9.0'

target 'Sample' do
  # Comment the next line if you don't want to use dynamic frameworks
  use_frameworks!
  project 'Sample/Sample.xcodeproj'
  # Pods for Sample
  pod 'Alamofire', '~> 5.5'
end
```

* Added a _container_ interface which is implemented by _projectContainer_ (`.xcodeproj`), _workspaceContainer_ (`.xcworkspace`); and _podWorkpace_ (an `.xcworkspace` where the actual workspace is not present, only the project files).

Resolves: https://bitrise.atlassian.net/browse/STEP-1804

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -6,16 +6,13 @@ require (
 	github.com/beevik/etree v1.1.0
 	github.com/bitrise-io/bitrise v0.0.0-20210519130014-380842fb41c1
 	github.com/bitrise-io/envman v0.0.0-20210517135508-b2b4fe89eac5
-	github.com/bitrise-io/go-steputils v0.0.0-20210514150206-5b6261447e77
-	github.com/bitrise-io/go-utils v0.0.0-20210520073355-367fa34178f5
-	github.com/bitrise-io/go-xcode v0.0.0-20210520114025-bdbec53ee59f
+	github.com/bitrise-io/go-steputils v1.0.1
+	github.com/bitrise-io/go-utils v1.0.1
+	github.com/bitrise-io/go-xcode v1.0.3-0.20220222074440-0c316f01fa68
```
Comment (lpusok) [inline go.mod:11]: Will need to be updated, when  https://github.com/bitrise-io/go-xcode/pull/157 merged.

File: scanners/ios/podfile.go [hunk 2]
```diff
@@ -49,12 +48,15 @@ begin
...
+	if !targets
+		puts "#{{}.to_json}"
+		exit
+	end
 
 	target_project_map = {}
 	targets.each do |name, target_definition|
 		next unless target_definition.user_project_path
-		target_project_map[name] = target_definition.user_project_path
+		target_project_map[name] = File.expand_path(target_definition.user_project_path)
```
Comment (lpusok) [inline scanners/ios/podfile.go:59]: Return absolute path.

File: scanners/ios/podfile.go [hunk 3]
```diff
@@ -111,21 +113,21 @@ func (podfileParser podfileParser) shouldRaiseReadDefinitionError(err string) bo
...
 	targetProjectMap, err := podfileParser.getTargetDefinitionProjectMap(cocoapodsVersion)
 	if err != nil {
 		return "", fmt.Errorf("failed to get target definition map: %s", err)
 	}
 
-	for target, project := range targetProjectMap {
-		if target == "Pods" {
-			return project, nil
-		}
+	// Return the first custom project
```
Comment (lpusok) [inline scanners/ios/podfile.go:122]: Fixed/worked around a problem that only top-level (Pod target) projects were parsed from Podfiles. Now using the first project, disregarding which target is it in.

# Uncomment the next line to define a global platform for your project
# platform :ios, '9.0'

target 'Sample' do
  # Comment the next line if you don't want to use dynamic frameworks
  use_frameworks!
  project 'Sample/Sample.xcodeproj'
  # Pods for Sample
  pod 'Alamofire', '~> 5.5'
end

File: scanners/ios/icon.go [hunk 4]
```diff
@@ -12,54 +12,29 @@ import (
...
+		return xcodeproj.Target{}, fmt.Errorf("scheme (%s) does not contain app buildable reference in project (%s)", scheme.Name, project.Path)
 	}
 
-	// Search for the main target
-	mainTarget, found := targetByBlueprintID(project.Proj.Targets, blueprintID)
+	blueprintID := entry.BuildableReference.BlueprintIdentifier
+	mainTarget, found := project.Proj.Target(blueprintID)
 	if !found {
-		return nil, fmt.Errorf("no target found for blueprint ID (%s) project (%s)", blueprintID, project.Path)
+		return xcodeproj.Target{}, fmt.Errorf("no target found for blueprint ID (%s) project (%s)", blueprintID, project.Path)
```
Comment (godrei) [inline scanners/ios/icon.go:24]: `no target found for blueprint ID (%s) project (%s)` -> `no target found for blueprint ID (%s) in project (%s)`

File: scanners/ios/podfile.go [hunk 5]
```diff
@@ -148,18 +150,18 @@ begin
 	# For example: require_relative '../node_modules/react-native/scripts/react_native_pods'
 	Dir.chdir(File.dirname(podfile_path))
 	podfile = Pod::Podfile.from_file(podfile_path)
-	pth = podfile.workspace_path
-	puts "#{{ :data => pth }.to_json}"
+	if !podfile.workspace_path
+		puts "#{{ :data => "" }.to_json}"
```
Comment (godrei) [inline scanners/ios/podfile.go:154]: why not use the same as in the case of `getTargetDefinitionProjectMap`:

```
	if !targets
		puts "#{{}.to_json}"
		exit
	end
```

File: scanners/ios/podfile.go [hunk 6]
```diff
@@ -299,57 +304,52 @@ func (podfileParser podfileParser) fixPodfileQuotation(podfilePth string) error
...
 			if found {
-				return []xcodeproj.ProjectModel{}, []xcodeproj.WorkspaceModel{}, fmt.Errorf("pod workspace (%s) found, but assigned project (%s) marked as standalone", podWorkspaceFile, podProjectFile)
+				return containers{}, fmt.Errorf("pod workspace (%s) found, but assigned project (%s) marked as standalone", podWorkspaceFile, podProjectFile)
 			}
-
-			mergedStandaloneProjects = standaloneProjects
-			mergedWorkspaces = ReplaceWorkspaceInList(workspaces, podWorkspace)
 		} else {
 			// Workspace not found, this means workspace is not in the repository,
 			// but it will created by `pod install`.
```
Comment (godrei) [inline scanners/ios/podfile.go:336]: but it will be created by

File: scanners/ios/project.go [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+				count++
+			}
+		}
+	}
+
+	return count
+}
+
+func pathRelativeToWorkspace(project, workspace string) string {
+	parentDir, _ := filepath.Split(workspace)
```
Comment (godrei) [inline scanners/ios/project.go:155]: why not `filepath.Dir` ?

File: scanners/ios/project.go [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+	relPath, err := filepath.Rel(filepath.Join(parentDir), project)
```
Comment (godrei) [inline scanners/ios/project.go:156]: why is the `filepath.Join(parentDir)` needed?

File: scanners/ios/project.go [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+	if err != nil {
+		log.Warnf("%s", err)
+		return project
+	}
+
+	return relPath
+}
+
+func printSchemes(includeUserSchemes bool, containerToSchemes map[string][]xcscheme.Scheme, containerPath string) {
```
Comment (godrei) [inline scanners/ios/project.go:165]: it seems to be unused

Comment (godrei) [reply scanners/ios/project.go:165]: this means `pathRelativeToWorkspace` has a single usage, and it is used for logging, do we need this? can not we log the absolute path and spare some lines of code?

**Review by godrei — CHANGES_REQUESTED** [review]

File: scanners/ios/icon.go (hunk 4 above)
Comment (lpusok) [reply scanners/ios/icon.go]: fixed

File: scanners/ios/podfile.go (hunk 5 above)
Comment (lpusok) [reply scanners/ios/podfile.go]: updated

File: scanners/ios/project.go (hunk 7 above)
Comment (lpusok) [reply scanners/ios/project.go]: removed completely.

File: scanners/ios/project.go (hunk 9 above)
Comment (lpusok) [reply scanners/ios/project.go]: removed completely.

File: scanners/ios/project.go (hunk 8 above)
Comment (lpusok) [reply scanners/ios/project.go]: removed, seems like it was not intended this way.

**Review by godrei — APPROVED** [review]

**Pull request merged**